Prevent read procedures from writing

boggle opened the pull request:
This makes sure we inject a specialized GDS into procedures that controls which capabilities are available to them. To do so, it introduces a lightweight AccessLevel facility to KernelTransaction.
Additionally, the GDS instance injected into procedures cannot: open/close transactions, work with kernel hooks, and autoindexing at this time.

pontusmelke commented on community/kernel/src/main/java/org/neo4j/kernel/AvailabilityGuard.java:
```diff
@@ -25,6 +25,7 @@
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.function.Function;
 
+import org.neo4j.graphdb.DatabaseShutdownException;
```
used?

pontusmelke commented on community/kernel/src/main/java/org/neo4j/kernel/AvailabilityGuard.java:
```diff
@@ -72,6 +73,7 @@ public UnavailableException( String message )
         String description();
     }
 
+
```
why?

pontusmelke commented on community/kernel/src/main/java/org/neo4j/kernel/api/ReadOperations.java:
```diff
@@ -190,25 +201,25 @@ RelationshipIterator nodeGetRelationships( long nodeId,
      * @throws org.neo4j.kernel.api.exceptions.index.IndexNotFoundKernelException if no such index found.
      */
-    long nodeGetFromUniqueIndexSeek( IndexDescriptor index, Object value ) throws IndexNotFoundKernelException,
+    long nodeGetFromUniqueIndexSeek(IndexDescriptor index, Object value) throws IndexNotFoundKernelException,
```
I believe the old formatting was the _correct_ one? (true for all changes in this file)

pontusmelke commented on community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/ListLabelsProcedure.java:
```diff
@@ -42,7 +42,7 @@ public ListLabelsProcedure( ProcedureName name )
     public RawIterator<Object[], ProcedureException> apply( Context ctx, Object[] input ) throws ProcedureException
     {
-        RawIterator<Label,ProcedureException> labels = asRawIterator( TokenAccess.LABELS.inUse( ctx.get( statement ) ) );
+        RawIterator<Label,ProcedureException> labels = asRawIterator( TokenAccess.LABELS.inUse( ctx.get(STATEMENT) ) );
```
formatting: `ctx.get( STATEMENT )`

pontusmelke commented on community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/ListPropertyKeysProcedure.java:
```diff
@@ -41,7 +41,7 @@ public ListPropertyKeysProcedure( ProcedureSignature.ProcedureName name )
     public RawIterator<Object[], ProcedureException> apply( Context ctx, Object[] input ) throws ProcedureException
     {
-        RawIterator<String,ProcedureException> labels = asRawIterator( TokenAccess.PROPERTY_KEYS.inUse( ctx.get( statement ) ) );
+        RawIterator<String,ProcedureException> labels = asRawIterator( TokenAccess.PROPERTY_KEYS.inUse( ctx.get(STATEMENT) ) );
```
formatting

pontusmelke commented on community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/ListRelationshipTypesProcedure.java:
```diff
@@ -43,7 +43,7 @@ public ListRelationshipTypesProcedure( ProcedureSignature.ProcedureName name )
     {
         RawIterator<RelationshipType,ProcedureException> types =
-                asRawIterator( TokenAccess.RELATIONSHIP_TYPES.inUse( ctx.get( statement ) ) );
+                asRawIterator( TokenAccess.RELATIONSHIP_TYPES.inUse( ctx.get(STATEMENT) ) );
```
formatting

pontusmelke commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java:
```diff
@@ -1395,21 +1393,28 @@ public DoubleLongRegister indexSample( IndexDescriptor index, DoubleLongRegister
+            }
         }
-        finally
+        catch ( UnauthorizedAccessKernelException e )
```
Maybe get rid one level of nesting here? Turn the inner try catch to a method.

pontusmelke commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/StateHandlingStatementOperations.java:
```diff
@@ -66,6 +66,7 @@
 import org.neo4j.kernel.api.properties.PropertyKeyIdIterator;
 import org.neo4j.kernel.api.txstate.TransactionCountingStateVisitor;
 import org.neo4j.kernel.api.txstate.TransactionState;
+import org.neo4j.kernel.api.index.AutoIndexing;
```
remove

pontusmelke commented on community/neo4j-harness/neo4j-home/ssl/snakeoil.cert:
```diff
@@ -0,0 +1,11 @@
+-----BEGIN CERTIFICATE-----
```
remove

pontusmelke commented on community/neo4j-harness/neo4j-home/ssl/snakeoil.key:
```diff
@@ -0,0 +1,16 @@
+-----BEGIN PRIVATE KEY-----
```
remove

pontusmelke commented on community/server/neo4j-home/ssl/snakeoil.cert:
```diff
@@ -0,0 +1,11 @@
+-----BEGIN CERTIFICATE-----
```
remove

pontusmelke commented on community/server/neo4j-home/ssl/snakeoil.key:
```diff
@@ -0,0 +1,16 @@
+-----BEGIN PRIVATE KEY-----
```
remove

pontusmelke commented on community/kernel/src/main/java/org/neo4j/kernel/api/ReadOperations.java:
```diff
@@ -58,8 +58,7 @@
  */
-public interface ReadOperations
-{
+public interface ReadOperations {
```
This file is not formatted correctly

pontusmelke commented on community/server/data/dbms/auth:
```diff
@@ -0,0 +1 @@
+neo4j:SHA-256,02D1DC324C275D63801047F3B37E3D47C7BFCCD1DE607BE433C5F64A3BFC7D28,AEF1BB85ACFB248E7F1BCA5FA216F067:password_change_required
```
remove

MishaDemianenko commented on enterprise/core-edge/src/main/java/org/neo4j/coreedge/server/edge/EnterpriseEdgeEditionModule.java:
```diff
@@ -117,18 +118,14 @@ public EnterpriseEdgeEditionModule( final PlatformModule platformModule,
-            {
-            }
-        };
+        schemaWriteGuard = () -> {};
```
There 3 implementations like that in a code base, that should be extracted and class should be introduced instead

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/exceptions/UnauthorizedAccessKernelException.java:
```diff
@@ -0,0 +1,28 @@
+{
+    public UnauthorizedAccessKernelException( String message )
+    {
+        super( Status.Transaction.UnauthorizedAccess, (Throwable) null, message );
```
why do we need to puss  `(Throwable) null`? There is another constructor without `Throwable`

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/AccessLevel.java:
```diff
@@ -52,8 +55,17 @@ public boolean allowsSchemaWrites()
     public abstract boolean allowsSchemaWrites();
+    public abstract boolean mayLimitAccessTo( AccessLevel newMode );
+
+    public interface Reset
```
Does not sound like a good place for something that will reset level in KernelTransaction, looks strange and suspicious

MishaDemianenko commented:
guys please avoid reformatting in a PRs that brings something else inside, it's really hard to follow

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/ReadOperations.java:
```diff
@@ -561,15 +603,24 @@ DoubleLongRegister indexSample( IndexDescriptor index, DoubleLongRegister target
+    /**
+     * For managed procedures, this gives access to the current transaction.
+     */
+    CallableProcedure.Key<KernelTransaction> KERNEL_TRANSACTION = CallableProcedure.Key.key( "KernelTransaction",
```
i find it really strange that predefined Context keys defined elsewhere, can we move it there?

boggle replied:
+1

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/GraphDatabaseFacadeFactoryTest.java:
```diff
@@ -143,7 +146,8 @@ public LifeSupport createLife()
             protected EditionModule createEdition( PlatformModule platformModule )
             {
-                return null;
+                EditionModule edition = Mockito.mock( EditionModule.class, Answers.RETURNS_DEEP_STUBS.get() );
```
please use `Mockito.RETURNS_DEEP_STUBS` instead

boggle replied:
Is there a reason for that? I thought both are basically equivalent and public API

MishaDemianenko replied:
`Mockito` is recommended public facade class that should decrease number of mockito classes that user code interact with

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ProcedureGDSFactory.java:
```diff
@@ -0,0 +1,213 @@
+        KernelTransaction transaction = context.get( ReadOperations.KERNEL_TRANSACTION );
+
+        GraphDatabaseFacade facade = new GraphDatabaseFacade();
+        facade.init( config, new GraphDatabaseFacade.SPI()
```
this SPI implementation should be extracted into it's own class

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/index/AutoIndexing.java:
```diff
@@ -0,0 +1,26 @@
+ */
+package org.neo4j.kernel.api.index;
+
+public interface AutoIndexing
```
class level and method level java docs are missing

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/index/AutoIndexing.java:
```diff
@@ -0,0 +1,26 @@
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.api.index;
```
since autoindexes is legacy indexes that should be reflected in package name

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/index/AutoIndexOperations.java:
```diff
@@ -0,0 +1,104 @@
+import org.neo4j.kernel.api.exceptions.legacyindex.AutoIndexingKernelException;
+import org.neo4j.kernel.api.properties.Property;
+
+public interface AutoIndexOperations
```
class level and method level javadocs

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/index/AutoIndexOperations.java:
```diff
@@ -0,0 +1,104 @@
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.api.index;
```
package name should reflect legacy index stuff

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/InternalAutoIndexing.java:
```diff
@@ -0,0 +1,62 @@
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.api.index;
```
legacy package

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/InternalAutoIndexOperations.java:
```diff
@@ -0,0 +1,316 @@
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.api.index;
```
package name should be changed to include legacy notion

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/InternalAutoIndexOperations.java:
```diff
@@ -0,0 +1,316 @@
+                throws LegacyIndexNotFoundKernelException, EntityNotFoundException;
+    }
+
+    private AtomicReference<Set<String>> propertyKeysToInclude = new AtomicReference<>( Collections.emptySet() );
```
if we will add propertyKeysToInclude to constructor we should be able to make it just a plain set

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/InternalAutoIndexOperations.java:
```diff
@@ -0,0 +1,316 @@
+    }
+
+    // Trap door needed to keep this as an enum
+    void replacePropertyKeysToInclude( List<String> propertyKeysToIncludeNow )
```
as soon as propertyKeysToInclude will be included in constructor this method should go.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/InternalAutoIndexOperations.java:
```diff
@@ -0,0 +1,316 @@
+            }
+            catch ( TokenNotFoundException e )
+            {
+                // TODO: TokenNotFoundException was added before there was a kernel. It should be converted to a
```
all TODOs in this class should be addressed

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacade.java:
```diff
@@ -164,9 +163,11 @@
 
         URL validateURLAccess( URL url ) throws URLAccessValidationError;
+
+        String toFacadeString();
```
What is the purpose of this method and why it should be in `GraphDatabaseFacade`?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ProcedureGDSFactory.java:
```diff
@@ -0,0 +1,213 @@
+import java.util.Map;
+import java.util.function.Supplier;
+
+public class ProcedureGDSFactory implements ThrowingFunction<CallableProcedure.Context,GraphDatabaseService,ProcedureException>
```
Factory should implement `Factory` instead of `ThrowingFunction`

boggle replied:
Can't use that, 1) we have an argument per call to the factory and don't want to create it again and again (maybe we could do that) 2) we need to throw a checked exception, factory interface doesn't allow that

MishaDemianenko replied:
Then it looks like you can use ThrowingSupplier for all of that

MishaDemianenko commented on community/collections/src/main/java/org/neo4j/helpers/collection/Iterables.java:
```diff
@@ -33,6 +33,7 @@
 import java.util.function.Predicate;
 
 import org.neo4j.collection.RawIterator;
+import org.neo4j.function.ThrowingAction;
```
redundant import

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/InternalAutoIndexOperations.java:
```diff
@@ -0,0 +1,316 @@
+    @Override
+    public void startAutoIndexingProperty( String propName )
+    {
+        propertyKeysToInclude.getAndUpdate( current -> {
```
as soon as set will be released from atomicReference, this can be simply achieved by any concurrent set implementations if multi threaded access is required.

boggle replied:
Since these sets are updated very, very rarely, I believe for reading, using an immutable set obtained from an atomic reference is faster than using a concurrent set.

MishaDemianenko replied:
i was more concerned about code simplicity then speed here

MishaDemianenko commented:
@boggle should we also prevent people from starting new thread inside a procedures?

MishaDemianenko commented:
@boggle finish my first review round; couple of comments and also needs rebase

boggle commented:
On formatting: Sorry that happens whenever I'm still on the scala style and is mostly done by IDEA without me noticing. Would be awesome to have a code style file that works for both Java and scala.

boggle commented:
On AutoIndexing: This is really just refactoring; I'm happy to move things into the right places but it's out of scope of this PR to address old TODOs. Happy to have a look at least.

boggle commented:
On preventing threads: No, we want them to be able to do that. Also achieving that is hard, you'd have to mess with classloaders to get that done or perhaps the Java security model.

boggle committed:
Introduce and inject CoreAPIAvailabilityGuard

boggle committed:
Introduce ProcedureGDSFactory

boggle committed:
Factor out SPI class from ProcedureGDSFacadeSPI

boggle committed:
Use ProcedureGDSFactory to prevent procedures from invoking blocked/unsupported operations, like db.shutdown

o Also fixes bug where draining an internal iterator in a ProcedureExecutionResult forgot to call success()

MishaDemianenko commented:
On Thread prevention: was thinking about custom security manager and thread group policies, since think that will bring more value.

MishaDemianenko commented:
On Auto indexing: yep saw that, can we just move it to another PR then? or you should promise to do that in next PR)?

boggle committed:
Inject StoreDir into ProcedureGDSFactory

boggle commented:
I had to rebuild all the commits here and tried to address your feedback.

boggle commented:
@MishaDemianenko Ping
